graph_def_editor/function_graph.py

```diff
@@ -0,0 +1,351 @@
+    """Returns the FuncGraph associated to the given func_name if possible."""
+    outer_graph = graph
+    while graph is not None:
+      # pylint: disable=protected-access
+      func = graph._get_function(str(func_name))
+      if func is not None:
+        if hasattr(func, "graph"):
+          return func.graph
+        # `outer_graph` may not be the same as `ops.get_default_graph()` e.g.
+        # in the case of nested if ops or when the gradient is being computed
+        # from inside a Defun. We build the `func_graph` with `outer_graph`
+        # as its outer graph.
+        with outer_graph.as_default():
+          # This is a _DefinedFunction.
+          func_graph = (
+              function_def_to_graph.function_def_to_graph(func.definition))
+        if func_graph is not None:
+          return func_graph
+      if hasattr(graph, "outer_graph"):
+        graph = graph.outer_graph
+      else:
+        raise ValueError(
+            "Function {} does not exist in the graph.".format(func_name))
+
+  def to_function_graph_def(self, add_shapes=True):
+    # type: (bool) -> function_pb2.FunctionDef
+    """
+    Args:
+      add_shapes: If True, add the special "_output_shapes" attribute with
+        output shape information from this Node's output metadata.
+
+    Returns the `function_pb2.FunctionDef` serialization of this function's
+    graph in its current form.
+    """
+    ret = function_pb2.FunctionDef()
+    ret.CopyFrom(self._func_graph_def)
+    # Leave signature as is, but replace all node_defs
+    del ret.node_def[:]
+    ret.signature.CopyFrom(self._func_graph_def.signature)
+
+    input_args = [input_arg.name for input_arg in ret.signature.input_arg]
+
+    for op in self.nodes:
+      if op.op_type == _INPUT_DUMMY_OP_NAME:
+        continue
+
+      node_def = ret.node_def.add()
+      op.to_node_def(node_def, add_shapes)
+      unique_input_counter = Counter()
+
+      for i in range(len(op.inputs)):
+        (input_tensor_name, global_input_index_str) = (
+            op.inputs[i].name.split(":"))
+
+        global_input_index = int(global_input_index_str)
+        if input_tensor_name in input_args:
+          # don't add index for function args
+          node_def.input[i] = input_tensor_name
+        else:
+          input_op_output_args, input_op_output_has_number_attr = (
+              self._get_op_def_denormalized_outputs(op.inputs[i].op))
+          if (len(input_op_output_args) == 1 and
+              input_op_output_args[0].type_list_attr):
+            node_def.input[i] = (
+                input_tensor_name + ":" + input_op_output_args[0].name + ":" +
+                str(global_input_index))
+          else:
+            input_name = (
+                input_tensor_name + ":" +
+                input_op_output_args[global_input_index].name)
+            node_def.input[i] = (
+                input_name + ":" + str(unique_input_counter[input_name]))
```

Comment:
It looks like this code splits list-valued inputs among ops in an order that depends on the iteration order of `self.nodes`. Is that ordering guaranteed to be the correct one?